File: scripts/batch_signup_vote.py
import time
import os
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException, NoSuchElementException

# Configuration
TARGET_URL = "http://localhost:8083"
USER_COUNT = 14
BASE_EMAIL = "firstlast{}@example.com"
PASSWORD = "password123"

def run_batch_signup():
    print(f"[Test] Starting UI Automation for {USER_COUNT} users...")
    
    options = webdriver.ChromeOptions()
    # options.add_argument("--headless") 
    options.add_argument(f"--user-data-dir={os.path.join(os.getcwd(), 'chrome-data')}")
    options.add_argument("--disable-extensions")
    options.add_argument("--no-sandbox")
    options.add_argument("--disable-dev-shm-usage")
    
    try:
        
        # Force webdriver_manager to install to local folder
        os.environ['WDM_LOCAL'] = '1'
        os.environ['WDM_SSL_VERIFY'] = '0'

        from webdriver_manager.chrome import ChromeDriverManager
        from selenium.webdriver.chrome.service import Service as ChromeService

        # Install driver
        service = ChromeService(ChromeDriverManager().install())
        driver = webdriver.Chrome(service=service, options=options)
    except Exception as e:
        print(f"[Error] Failed to launch Chrome: {e}")
        print("Ensure you have Chrome installed.")
        return

    wait = WebDriverWait(driver, 10)

    for i in range(1, USER_COUNT + 1):
        email = BASE_EMAIL.format(i)
        first_name = f"First{i}"
        last_name = f"Last{i}"
        
        print(f"\n[User {i}/{USER_COUNT}] Processing {email}...")
        
        try:
            # 1. Go to URL
            driver.get(TARGET_URL)
            
            # Check for existing logout (if previous session stuck)
            try:
                logout_btns = driver.find_elements(By.XPATH, "//div[contains(text(), 'Logout')]")
                if logout_btns:
                    logout_btns[0].click()
                    time.sleep(1)
            except:
                pass

            # 2. Navigate to Sign Up
            # Wait for email input to know page loaded
            wait.until(EC.presence_of_element_located((By.CSS_SELECTOR, "input[placeholder='Email']")))
            
            # Click "Sign Up" toggle if present (it's a div text)
            # We assume we are on Login by default.
            # Find the toggle that says "Sign Up"
            try:
                toggles = driver.find_elements(By.TAG_NAME, "div")
                for t in toggles:
                    if t.text == "Sign Up":
                        t.click()
                        break
            except:
                pass
            
            # 3. Fill Form
            wait.until(EC.presence_of_element_located((By.CSS_SELECTOR, "input[placeholder='First Name']"))).send_keys(first_name)
            driver.find_element(By.CSS_SELECTOR, "input[placeholder='Last Name']").send_keys(last_name)
            driver.find_element(By.CSS_SELECTOR, "input[placeholder='Email']").send_keys(email)
            driver.find_element(By.CSS_SELECTOR, "input[placeholder='Password']").send_keys(PASSWORD)
            
            # 4. Submit
            # Find the submit button (div/button with text "Sign Up")
            # Usually the last one or by style. Let's look for exact text match.
            submit_btns = driver.find_elements(By.XPATH, "//div[@role='button']//div[contains(text(), 'Sign Up')]")
            # Parent of that div
            if submit_btns:
                 # Go up to the button role
                 submit_btns[0].find_element(By.XPATH, "./..").click()
            else:
                 # Fallback for direct button
                 driver.find_element(By.XPATH, "//div[@role='button'][descendant::text()='Sign Up']").click()
            
            # 5. Wait for Home Screen
            # Look for "TAP TO JOIN" or "GameID"
            wait.until(EC.presence_of_element_located((By.XPATH, "//*[contains(text(), 'TAP TO JOIN') or contains(text(), 'GameID')]")))
            print("  - Signed up/Logged in.")

            # 6. Vote
            try:
                vote_btn = wait.until(EC.element_to_be_clickable((By.XPATH, "//div[contains(text(), 'TAP TO JOIN')]")))
                vote_btn.click()
                print("  - Clicked Vote.")
                
                # Handle Alert if any
                try:
                    WebDriverWait(driver, 3).until(EC.alert_is_present())
                    alert = driver.switch_to.alert
                    alert.accept()
                    print("  - Accepted Alert.")
                except TimeoutException:
                    pass
                    
                time.sleep(2) # Wait for write
            except TimeoutException:
                 print("  - 'TAP TO JOIN' not found (maybe already voted).")

            # 7. Logout
            try:
                logout = wait.until(EC.element_to_be_clickable((By.XPATH, "//div[contains(text(), 'Logout')]")))
                logout.click()
                print("  - Logged out.")
                time.sleep(1)
            except:
                print("  - Logout failed.")

        except Exception as e:
            print(f"  [!] Failed: {e}")
            
    print("\n[Test] Automation Complete.")
    driver.quit()

if __name__ == "__main__":
    run_batch_signup()

File: scripts/batch_signup_vote.js
const puppeteer = require('puppeteer-core');

(async () => {
    const TARGET_URL = 'http://localhost:8083'; // Ensure this matches your running server
    const USER_COUNT = 14;
    const BASE_NAME = 'firstlast';
    const PASSWORD = 'password123';

    console.log(`[Test] Starting UI Automation for ${USER_COUNT} users...`);

    // Launch browser
    const browser = await puppeteer.launch({
        headless: false,
        defaultViewport: null,
        args: ['--window-size=1280,800'],
        executablePath: '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome' // Point to system Chrome
    });

    const page = await browser.newPage();

    for (let i = 1; i <= USER_COUNT; i++) {
        const email = `${BASE_NAME}${i}@example.com`;
        const firstName = `First${i}`;
        const lastName = `Last${i}`;

        console.log(`\n[User ${i}/${USER_COUNT}] Processing ${email}...`);

        try {
            // 1. Go to Login/Signup Page
            await page.goto(TARGET_URL);

            // Check if we are already logged in (if previous logout failed)
            const logoutBtn = await page.$('text/Logout');
            if (logoutBtn) {
                console.log('  - Found residual session, logging out...');
                await logoutBtn.click();
                await page.waitForNavigation();
            }

            // 2. Navigate to Sign Up
            // Wait for "Don't have an account? Sign Up" button or link
            // Adjust selector based on your valid UI. currently it's "Sign Up" text text
            const signUpLink = await page.waitForSelector('div[role="button"]', { timeout: 2000 }).catch(() => null);
            // Actually, we need to find the specific "Sign Up" toggle text or button.
            // Let's assume the user starts at Login.

            // Wait for input fields to ensure page load
            await page.waitForSelector('input[placeholder="Email"]');

            // Click "Sign Up" toggle if we are in Login mode
            // We can look for text "Sign Up"
            const handles = await page.$$('div');
            let signUpClicked = false;
            for (const h of handles) {
                const text = await page.evaluate(el => el.textContent, h);
                if (text === 'Sign Up') {
                    await h.click();
                    signUpClicked = true;
                    break;
                }
            }

            if (!signUpClicked) {
                // Maybe it's a specific class? Let's try filling first.
            }

            // Fill Sign Up Form
            // We need to type into inputs.
            // Inputs: First Name, Last Name, Email, Password
            await page.type('input[placeholder="First Name"]', firstName);
            await page.type('input[placeholder="Last Name"]', lastName);
            await page.type('input[placeholder="Email"]', email);
            await page.type('input[placeholder="Password"]', PASSWORD);

            // 3. Submit
            // Find "Sign Up" button. It likely says "Sign Up" and is different from the toggle.
            // The button usually has a background color.
            const buttons = await page.$$('div[role="button"]');
            for (const b of buttons) {
                const t = await page.evaluate(el => el.textContent, b);
                if (t === 'Sign Up') {
                    await b.click();
                    break;
                }
            }

            // 4. Wait for Home Screen (redirect)
            // Look for "TAP TO JOIN" or "Next Game"
            await page.waitForFunction(
                () => document.body.textContent.includes('TAP TO JOIN') || document.body.textContent.includes('VOTED'),
                { timeout: 10000 }
            );
            console.log('  - Signed up successfully.');

            // 5. Vote
            // Click "TAP TO JOIN"
            // It might already be voted if we are re-using accounts, but these are new.
            const voteBtn = await page.$x("//div[contains(text(), 'TAP TO JOIN')]");
            if (voteBtn.length > 0) {
                await voteBtn[0].click();
                console.log('  - Clicked Vote button.');

                // Wait for status to change to "VOTED" or "JOINED"
                // Or wait for an alert? The app uses Alert.alert which might block puppeteer?
                // React Native Web Alert usually uses window.alert?
                // If it uses window.alert, we need to handle dialog.

                page.on('dialog', async dialog => {
                    // console.log('  - Dialog:', dialog.message());
                    await dialog.accept();
                });

                await new Promise(r => setTimeout(r, 2000)); // Wait a bit for processing
            } else {
                console.log('  - Vote button not found (already voted?)');
            }

            // 6. Logout
            // Try different selectors for logout
            let logout = await page.$x("//div[contains(text(), 'Logout')]");
            if (logout.length === 0) {
                logout = await page.$x("//button[contains(text(), 'Logout')]");
            }
            if (logout.length > 0) {
                await logout[0].click();
                console.log('  - Logged out.');
                await new Promise(r => setTimeout(r, 1000));
            } else {
                console.log('  - Logout button not found. User might be stuck logged in.');
            }
        } catch (e) {
            console.error(`  [!] Failed for user ${i}:`, e.message);
            // Try to recover by going back to base url
            await page.goto(TARGET_URL);
        }
    }

    console.log('\n[Test] Automation Complete.');
    await browser.close();
})();
